src/main/java/net/fabricmc/loader/impl/metadata/MetadataVerifier.java

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright 2016 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.loader.impl.metadata;
+
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import net.fabricmc.loader.api.FabricLoader;
+import net.fabricmc.loader.api.SemanticVersion;
+import net.fabricmc.loader.api.VersionParsingException;
+import net.fabricmc.loader.impl.FabricLoaderImpl;
+import net.fabricmc.loader.impl.discovery.ModCandidate;
+import net.fabricmc.loader.impl.util.log.Log;
+import net.fabricmc.loader.impl.util.log.LogCategory;
+
+public final class MetadataVerifier {
+	private static final Pattern MOD_ID_PATTERN = Pattern.compile("[a-z][a-z0-9-_]{1,63}");
+
+	public static ModCandidate verifyIndev(ModCandidate mod) {
+		if (FabricLoaderImpl.INSTANCE.isDevelopmentEnvironment()) {
+			try {
+				MetadataVerifier.verify(mod.getMetadata());
+			} catch (ParseMetadataException e) {
+				e.setModPaths(mod.getLocalPath(), Collections.emptyList());
+				throw new RuntimeException("Invalid mod metadata", e);
+			}
+		}
+
+		return mod;
+	}
+
+	static void verify(LoaderModMetadata metadata) throws ParseMetadataException {
+		checkModId(metadata.getId(), "mod id");
+
+		for (String providesDecl : metadata.getProvides()) {
+			checkModId(providesDecl, "provides declaration");
+		}
+
+		// TODO: verify mod id and version decls in deps
+
+		if (FabricLoader.getInstance().isDevelopmentEnvironment()) {
```

Comment:
Nit: is this line needed? as it checks if its a dev env in the above method.

Reply:
there are some direct calls to verify from the mod discoverer, the public verifyIndev is only used for builtin mods

Reply:
Ah yeah, my bad 👍